File: 200-number-of-islands/200-number-of-islands.cpp
class Solution {
public:
    
    void backtrack(vector <vector<char>>& grid, int r, int c, int rows, int cols)
    {
        if(r < 0 || r >= rows || c < 0 || c >= cols || grid[r][c] == '0')
            return;
        grid[r][c] = '0';
        backtrack(grid, r-1, c, rows, cols);
        backtrack(grid, r+1, c, rows, cols);
        backtrack(grid, r, c + 1, rows, cols);
        backtrack(grid, r, c - 1, rows, cols);
        
    }
    
    int numIslands(vector<vector<char>>& grid) 
    {
        int count = 0;
        int rows = grid.size(); int cols = grid[0].size();
        for(int r = 0; r < rows; r++)
        {
            for(int c = 0; c < cols; c++)
            {
                if(grid[r][c] == '1')
                {
                    count++;
                    backtrack(grid, r, c, rows, cols);
                }
            }
        }
        
        return count;
    }
};

File: 200-number-of-islands/200-number-of-islands.py
class Solution:
    
    def __init__(self):
        self.visited = set()
        self.queue = deque()
    
    def bfs(self, grid, rows, columns, r, c):
        
        self.queue = deque([[r, c]])
        while self.queue:
            r,c = self.queue.popleft()
            if (r, c) not in self.visited and grid[r][c] == '1':
                self.visited.add((r,c))
                
                if r != 0 and grid[r-1][c] == '1':
                    self.queue.append([r-1, c])
                if c!= 0 and grid[r][c-1] == '1':
                    self.queue.append([r, c-1])
                if r!=rows-1 and grid[r+1][c] == '1':
                    self.queue.append([r+1, c])
                if c!=columns - 1 and grid[r][c+1] == '1':
                    self.queue.append([r, c+1])
        
    def numIslands(self, grid: List[List[str]]) -> int:
        
        rows, columns = len(grid), len(grid[0])
        count = 0
        for r in range(rows):
            for c in range(columns):
                
                if grid[r][c] == "1" and (r, c) not in self.visited:
                    self.bfs(grid, rows, columns, r, c)
                    count += 1
        return count